packages/utils/src/queue/index.ts

```diff
@@ -0,0 +1,377 @@
+      this.queue, job,
+      (a: Readonly< Job<JobOptions, JobReturnType>>, b: Readonly< Job<JobOptions, JobReturnType>>) => b.priority - a.priority
+    )
+    this.queue.splice(index, 0, job)
+  }
+
+  /**
+   * Adds a sync or async task to the queue. Always returns a promise.
+   */
+  async add (fn: RunFunction<JobOptions, JobReturnType>, options?: JobOptions): Promise<JobReturnType> {
+    options?.signal?.throwIfAborted()
+
+    const job = new Job<JobOptions, JobReturnType>(fn, options, options?.priority)
+
+    const p = job.join(options)
+      .then(result => {
+        this.safeDispatchEvent('completed', { detail: result })
+
+        return result
+      })
+      .catch(err => {
+        this.safeDispatchEvent('error', { detail: err })
+
+        throw err
+      })
+
+    this.enqueue(job)
+    this.safeDispatchEvent('add')
+    this.tryToStartAnother()
+
+    return p
+  }
+
+  /**
+   * Clear the queue
+   */
+  clear (): void {
+    this.queue.splice(0, this.queue.length)
+  }
+
+  /**
+   * Abort all jobs in the queue and clear it
+   */
+  abort (): void {
+    this.queue.forEach(job => {
+      job.abort(new AbortError())
+    })
+
+    this.clear()
+  }
+
+  /**
+   * Can be called multiple times. Useful if you for example add additional items at a later time.
+   *
+   * @returns A promise that settles when the queue becomes empty.
+   */
+  async onEmpty (): Promise<void> {
+    // Instantly resolve if the queue is empty
+    if (this.size === 0) {
+      return
+    }
+
+    await this.onEvent('empty')
+  }
+
+  /**
+   * @returns A promise that settles when the queue size is less than the given
+   * limit: `queue.size < limit`.
+   *
+   * If you want to avoid having the queue grow beyond a certain size you can
+   * `await queue.onSizeLessThan()` before adding a new item.
+   *
+   * Note that this only limits the number of items waiting to start. There
+   * could still be up to `concurrency` jobs already running that this call does
+   * not include in its calculation.
+   */
+  async onSizeLessThan (limit: number): Promise<void> {
+    // Instantly resolve if the queue is empty.
+    if (this.size < limit) {
+      return
+    }
+
+    await this.onEvent('next', () => this.size < limit)
+  }
+
+  /**
+   * The difference with `.onEmpty` is that `.onIdle` guarantees that all work
+   * from the queue has finished. `.onEmpty` merely signals that the queue is
+   * empty, but it could mean that some promises haven't completed yet.
+   *
+   * @returns A promise that settles when the queue becomes empty, and all
+   * promises have completed; `queue.size === 0 && queue.pending === 0`.
+   */
+  async onIdle (): Promise<void> {
+    // Instantly resolve if none pending and if nothing else is queued
+    if (this.pending === 0 && this.size === 0) {
+      return
+    }
+
+    await this.onEvent('idle')
+  }
+
+  private async onEvent (event: keyof QueueEvents<JobReturnType>, filter?: () => boolean): Promise<void> {
```

Comment:
Does this need to be abortable?
I guess not if clear/abort will allow all pending promises (from public methods, onEmpty, onSizeLessThan, onIdle, toGenerator) to resolve.

Reply:
It's not abortable in p-queue, though it makes sense

Reply:
They are abortable now